Update installation guide to match how-to template

DragaDoncila opened the pull request:
This PR updates the installation guide to match our new docs template:
- brings prerequisites to the top
- highlights that we strongly recommend installing as a package
- adds admonitions in relevant spots to make it easier to browse
- adds dropdowns for the three different package install options to make it easier to browse
- tidies up some wording
- consolidates links and suggestions in the "next steps" section

DragaDoncila committed:
Update installation guide to match how-to template

GenevieveBuckley commented on tutorials/fundamentals/installation.md:
```diff
@@ -1,30 +1,47 @@
-This tutorial will teach you how to do a clean install of **napari**. It is
-aimed at people that just want to use napari. For people also interested in
-contributing to napari please check our [contributing
-guidelines](https://github.com/napari/napari/blob/master/docs/developers/CONTRIBUTING.md)
```
I think we may want to keep this redirection for people looking for the developer installation guide instead

GenevieveBuckley replied:
Ah, that link doesn't work anyway. Here it is: https://napari.org/developers/contributing.html

DragaDoncila replied:
Yep, I've added this at the very top in a note:

```{note} 
If you want to contribute back to the napari codebase, you should follow [this guide instead](https://napari.org/developers/contributing.html).
```

GenevieveBuckley commented on tutorials/fundamentals/installation.md:
```diff
@@ -1,30 +1,47 @@
+for installing the bundled app.
 
-## installation
+### Installing as a python package or in editable mode
```
Can we flag this as the method as "recommended"?

The bundle is still a bit experimental, and it's not so easy for people to add new packages to the bundle environment, etc. I think generally we prefer if people use napari installed as a python package.

GenevieveBuckley commented on tutorials/fundamentals/installation.md:
```diff
@@ -1,30 +1,47 @@
+This is the easiest way to install napari if you only wish to use it as a standalone GUI app.
+This installation method does not have any prerequisites. 
+[Click here](#installing-as-a-bundled-app) to see instructions
+for installing the bundled app.
```
Possibly we should talk about the restrictions involved with choosing the bundled app over the other methods

DragaDoncila replied:
@GenevieveBuckley I added an important note above:

```{important}
Note that the bundled app is still
in active development, and may not be very stable. We strongly recommend installing as a Python
package instead.
```

Is this sufficient mention of the restrictions?

GenevieveBuckley replied:
Yes, I think so. Might be good to link to the python installation section/instructions in that sentence, too.

GenevieveBuckley commented on tutorials/fundamentals/installation.md:
```diff
@@ -233,30 +250,25 @@ menu](../assets/tutorials/installation/windows_launch_napari.png)
 below for how to contact us.)
 
-## bug reports
+***
```
What's this line about? Was there supposed to be bold text here?

DragaDoncila replied:
this was to include a horizontal line separator between the content headings and the template headings, but I think stuff like that is supposed to be dictated by the theme instead.

GenevieveBuckley commented on tutorials/fundamentals/installation.md:
````diff
@@ -58,15 +75,15 @@ conda-forge channel. You can install it with:
 ```
 
-### clone the repository locally and install in editable mode
+### Clone the repository locally and install in editable mode
````
I don't think this content should live here - editable pip installations are for the developer docs (and these instructions are also slightly different to what's written there https://napari.org/developers/contributing.html)

GenevieveBuckley commented on tutorials/fundamentals/installation.md:
```diff
@@ -58,15 +75,15 @@ conda-forge channel. You can install it with:
-### install from the master branch on Github
+### Install from the master branch on Github
 
 To install the "next-release" version from github via pip, call
```
```suggestion
To install the "bleeding edge" version from github via pip, call
```

GenevieveBuckley replied:
I don't think "next-release" is very clear (but "bleeding edge" is also not very clear, so if you have any better ideas we should do that)

melissawm replied:
Development version? Unstable version?

GenevieveBuckley commented:
I think some of my comments were more about content, rather than structure, which might not be what you're looking for on this PR.

melissawm commented on tutorials/fundamentals/installation.md:
```diff
@@ -1,30 +1,47 @@
+
+***
+
+### Choose which distribution to install.
```
```suggestion
## Choose which distribution to install.
```
Going up one level ensures people understand the next sections are options and they should only go with one of them. I'd go as far as using numbering to make that even clearer:

1. From pip
2. From conda-forge
3. From the main branch on GitHub

DragaDoncila replied:
Yeah I think it would be nice if this was a dropdown selection similar to [this section](https://numpy.org/numpy-tutorials/content/pairing.html#pair-your-notebook-files-ipynb-and-md) in the numpy notebook/jupytext guide:

1. It would limit the amount of code those who want to install the bundled app see at first glance if scrolling through the document
2. It would limit distractions for those who just want to install via pip or via conda - you probably don't want to do all of them and if you do, it doesn't feel like that much effort to click all three
3. It clearly shows that these are three separate options and in general you would pick just one.

Thoughts? I'm not sure how easy this would be to implement given the current theme because I don't remember seeing any drop down/accordion behaviour. Pinging @codemonkey800 - do we have something like this? If not, could we bring over something like the [FAQ dropdown](https://www.napari-hub.org/faq) from the hub?

melissawm replied:
That is a [jupyter-book feature](https://jupyterbook.org/interactive/hiding.html) so might be already doable - although I don't know about the styling using the napari theme.

codemonkey800 replied:
The napari tutorials use this feature extensively, so we'll also have support for it in the napari theme 😄 

You can see a demo of it in #179.

melissawm commented on tutorials/fundamentals/installation.md:
````diff
@@ -58,15 +75,15 @@ conda-forge channel. You can install it with:
 ```
 
-### install from the master branch on Github
+### Install from the master branch on Github
````
```suggestion
### Install from the main branch on Github
```

DragaDoncila commented:
> I think some of my comments were more about content, rather than structure, which might not be what you're looking for on this PR.

@GenevieveBuckley we may as well do it while we're here. I don't think any of this intersects with the "Absolute Beginners" document (I feel like we would be linking to that document from here), so I think we should make the improvements you suggest to this document.

melissawm commented:
When this is merged, can we remove the duplicate installation instructions from the site home page?

GenevieveBuckley commented:
> When this is merged, can we remove the duplicate installation instructions from the site home page?

Could you link to both places? I agree that we probably don't want duplication, that's harder to maintain.

DragaDoncila committed:
Updates after review

DragaDoncila committed:
Merge branch 'main' into reformat-installation

DragaDoncila committed:
Rearrange and test with existing theme

DragaDoncila committed:
Merge branch 'main' into reformat-installation

DragaDoncila commented:
@GenevieveBuckley @melissawm I think this is ready for another round of reviews. I've converted it to MyST markdown so we can take advantage of all the new theme options, and I think I've addressed the feedback I've received. I tried to clean up the sections a bit too, and I've put the three different package install options into dropdowns.

I haven't yet managed to build it with the new theme, I'll make sure I do that before we merge, but content wise I'd love another pass from you both.

DragaDoncila committed:
Small updates after previewing with the new theme

GenevieveBuckley commented on tutorials/fundamentals/installation.md:
```diff
@@ -233,30 +275,24 @@ menu](../assets/tutorials/installation/windows_launch_napari.png)
-## bug reports
-
-If you are running into issues, please open a new issue on our [issue
-tracker](https://github.com/napari/napari/issues) and include the output of the
```
Did we lose this content on getting help if needed, or has it just been moved elsewhere in the document? I think the link for where to submit an issue is useful information.

DragaDoncila replied:
this has been moved to the bottom under "Next steps".

GenevieveBuckley commented:
I don't have additional comments, this seems mostly fine to me. Nothing much new from me.

There are a few places where I have an outstanding comment or two about not wanting to lose information that was included in this document before (like redirecting people to the developer install guide if they need that instead, how to file an issue if things went wrong with the install, etc.)

melissawm commented on tutorials/fundamentals/installation.md:
```diff
@@ -1,55 +1,106 @@
-# napari installation tutorial
+---
+jupyter:
+  jupytext:
+    formats: ipynb,md
+    text_representation:
+      extension: .md
+      format_name: markdown
+      format_version: '1.3'
+      jupytext_version: 1.13.0
+  kernelspec:
+    display_name: Python 3 (ipykernel)
+    language: python
+    name: python3
+---
```
Do we want to make this document executable?

DragaDoncila replied:
I've made this document executable for now purely so I can use the Jupyter book dropdown feature, and also because I think in general we want all tutorials to be MyST markdown eventually. 
Once we merge the dropdown accordion feature (issue #279) we can change this dropdown to match that styling. Do you think there are any dangers to keeping it executable until then @melissawm?

melissawm commented on tutorials/fundamentals/installation.md:
```diff
@@ -1,55 +1,106 @@
 
-napari can be installed on most macOS, Linux, and Windows systems with Python
-3.7, 3.8, and 3.9 using pip:
+## Choose which distribution to install
```
I feel like we have redundant information here, since the different installation methods have already been mentioned above. Maybe the "Choose which distribution to install" section could come before the particular installation methods.

DragaDoncila replied:
Yeahh this one is a bit weird because "Choose which distribution to install" does only refer to installing as package - it's just do you want it from `pip`, `conda` or `github main branch`. I'll change the wording here to reflect this a bit better.

melissawm reviewed (commented):
Thanks @DragaDoncila , I only have one observation about the content, overall looks good (and once this is merged I would suggest cleaning up the index page!)

DragaDoncila committed:
Reword some sentences and rearrange to make the python package options clearer

DragaDoncila committed:
merge main

DragaDoncila committed:
Tweak workding agian

DragaDoncila commented:
Ok so @GenevieveBuckley and @melissawm I actually don't need to move this to `napari/napari` since tutorials are in `napari.github.io` anyway. 

I'd love if either one of you could do a final pass on this one. I think I've addressed all the outstanding comments? I've replied to each comment I think.

I'll also add a couple of screenshots here to show what it looks like with the theme:
![Screen Shot 2021-11-30 at 11 47 14 AM](https://user-images.githubusercontent.com/17995243/143968850-f4651c4e-e3ee-467c-8653-503b6a6f225f.png)
![Screen Shot 2021-11-30 at 11 47 42 AM](https://user-images.githubusercontent.com/17995243/143968860-363fe0ae-73ae-4b50-be58-4bc5729ca91e.png)

It might also be good to get a review from someone outside the docs working group. @sofroniewn @tlambert03?

sofroniewn reviewed (approved):
LGTM! Thanks for updating!!

GenevieveBuckley commented on tutorials/fundamentals/installation.md:
````diff
@@ -1,55 +1,101 @@
+This guide will teach you how to do a clean install of **napari** and launch the viewer.
+
+```{note} 
+If you want to contribute back to the napari codebase, you should follow [this guide instead](https://napari.org/developers/contributing.html).
````
```suggestion
If you want to contribute code back into napari, you should follow the [development installation instructions in the contributing guide](https://napari.org/developers/contributing.html) instead.
```

I'd kind of like the words "development installation" included in this sentence, otherwise it's a bit hard to see how this page content relates to this note.

GenevieveBuckley commented on tutorials/fundamentals/installation.md:
```diff
@@ -1,55 +1,101 @@
 
-### Which distribution to install
+It requires:
+- [Python 3.7+](https://www.python.org/downloads/)
```
```suggestion
- [Python 3.7 or higher](https://www.python.org/downloads/)
```

Just for clarity :smile:

GenevieveBuckley replied:
Although, I'm not sure you should link to https://www.python.org/downloads/ if you are later going to try to encourage the use of conda. Maybe the link can be removed?

GenevieveBuckley commented on tutorials/fundamentals/installation.md:
`````diff
@@ -1,55 +1,101 @@
+Python package distributions of napari can be installed via `pip`, `conda-forge`, or from source.
 
 ````{important}
 While not strictly required, it is *highly* recommended to install
`````
```suggestion
While not strictly required, it is highly recommended to install
```

Personal preference, I'd prefer we didn't have such a mix of emphasis markers so close together. (There is bold text above for **(Highly recommended)**, the "Important" box is also for emphasis, I don't think we need a single word in italics to further emphasize things on top of all that)

GenevieveBuckley commented on tutorials/fundamentals/installation.md:
````diff
@@ -1,55 +1,101 @@
+3.7, 3.8, and 3.9 using pip:
+
+```sh
 pip install "napari[all]"
````
```suggestion
python -m pip install "napari[all]"
```

Suggestion: we should consider making it a general rule that anytime we have instructions with pip, we show `python -m pip ...`. It reduces the chance that someone unfamiliar with virtual environments will end up pip installing packages into their base environment instead by mistake, etc.

GenevieveBuckley commented on tutorials/fundamentals/installation.md:
````diff
@@ -1,55 +1,101 @@
+You can then upgrade napari to a new version using
+
+```sh
+pip install "napari[all]" --upgrade
````
```suggestion
python -m pip install "napari[all]" --upgrade
```

GenevieveBuckley commented on tutorials/fundamentals/installation.md:
`````diff
@@ -1,55 +1,101 @@
 ````{important}
 While not strictly required, it is *highly* recommended to install
 napari into a clean virtual environment using an environment manager like
 [conda](https://docs.conda.io/en/latest/miniconda.html) or
`````
This is my favourite docs page on how to use conda environments, maybe it's worthwhile linking to it here? Up to you.
https://docs.conda.io/projects/conda/en/latest/user-guide/tasks/manage-environments.html

DragaDoncila replied:
Lovely, thanks Gen!

GenevieveBuckley commented on tutorials/fundamentals/installation.md:
```diff
@@ -58,24 +104,25 @@ conda-forge channel. You can install it with:
-git clone https://github.com/napari/napari.git
-cd napari
-pip install -e .[all]
+pip install "git+https://github.com/napari/napari.git#egg=napari[all]"
```
```suggestion
python -m pip install "git+https://github.com/napari/napari.git#egg=napari[all]"
```

GenevieveBuckley reviewed (approved):
Looking great, @DragaDoncila 

I've left some minor comments, but I think we can merge this in the next day or so. Good work!

DragaDoncila committed:
Update tutorials/fundamentals/installation.md

Co-authored-by: Genevieve Buckley <30920819+GenevieveBuckley@users.noreply.github.com>

DragaDoncila committed:
Update tutorials/fundamentals/installation.md

Co-authored-by: Genevieve Buckley <30920819+GenevieveBuckley@users.noreply.github.com>

DragaDoncila committed:
Update tutorials/fundamentals/installation.md

Co-authored-by: Genevieve Buckley <30920819+GenevieveBuckley@users.noreply.github.com>

DragaDoncila committed:
Update tutorials/fundamentals/installation.md

Co-authored-by: Genevieve Buckley <30920819+GenevieveBuckley@users.noreply.github.com>

DragaDoncila committed:
Update tutorials/fundamentals/installation.md

Co-authored-by: Genevieve Buckley <30920819+GenevieveBuckley@users.noreply.github.com>

DragaDoncila committed:
Update tutorials/fundamentals/installation.md

Co-authored-by: Genevieve Buckley <30920819+GenevieveBuckley@users.noreply.github.com>

DragaDoncila committed:
Add link to managing conda envs

DragaDoncila commented:
@GenevieveBuckley thanks for that final round of very thoughtful comments! I've updated and I think this is ready now.

GenevieveBuckley commented on tutorials/fundamentals/installation.md:
```diff
@@ -1,55 +1,101 @@
+Prerequisites differ depending on how you want to install napari.
 
-## Installation
+### Installing as a Python package 
```
```suggestion
### Prerequisites for installing napari as a Python package 
```

GenevieveBuckley commented on tutorials/fundamentals/installation.md:
```diff
@@ -1,55 +1,101 @@
-If you only wish to use napari as GUI app (and not from python code), the
-[bundled app](#installing-as-a-bundled-app)
-is the easiest way to install, it does not require having python pre-installed.
+### Installing the bundled app
```
```suggestion
### Prerequisites for installing napari as a bundled app
```

GenevieveBuckley commented on tutorials/fundamentals/installation.md:
```diff
@@ -1,55 +1,101 @@
 
-napari can be installed on most macOS, Linux, and Windows systems with Python
-3.7, 3.8, and 3.9 using pip:
+## Install as Python package
```
```suggestion
## Install as Python package (recommended)
```

GenevieveBuckley committed:
Update tutorials/fundamentals/installation.md

GenevieveBuckley committed:
Update tutorials/fundamentals/installation.md

GenevieveBuckley committed:
Update tutorials/fundamentals/installation.md

GenevieveBuckley commented:
Looks good!

I've just updated some of the headings to make it clearer which are part of the prerequisites section and which are part of the instructions section, and will merge the PR now.

Thanks for working on this @DragaDoncila
